### PR TITLE
[Core][Frontend] Add Support for Inference Time mm_processor_kwargs

### LINKED ISSUE #8742: [Feature]: Support Inference Overrides for mm_processor_kwargs
### 🚀 The feature, motivation and pitch

Follow-up on https://github.com/vllm-project/vllm/pull/8657, which added support for passing initialization time `mm_processor_kwargs` to be used by the input mapper / processor / max token count calculations / dummy data if they're added to architecture-specific implementations as keyword arguments. It would be nice to also be to pass such kwargs as input values at inference time as part of the multi-modal data, e.g.,:

```python
llm.generate({"multi_modal_data": {"image": {"data": image, "mm_processor_kwargs": image_kwargs}}})
```

Such that for models that support additional `mm_processor_kwargs`:
- [ ] The initialization time `mm_processor_kwargs` take priority over the config values
- [ ] The inference time `mm_processor_kwargs` take priority over the config values and the initialization `mm_processor_kwargs`

### Alternatives

Keep `mm_processor_kwargs` as initialization time only

### Additional context

For per-request `mm_processor_kwargs`, it needs to be correctly handled:

- In the input mapper
- In the input processor

Some care needs to be taken around the input mapper, which falls back to a wrapper around HF resources, e.g., image processors, since it may take stuff out of the config. More specifically:

- We should avoid initializing and managing multiple multimodal processors with different processor kwargs if possible
- Init time processor kwargs / per request processor kwargs should behave identically - this probably depends on the `preprocess` signature for the HF resource closely matching the `init` signature by default
    - If for whatever reason init/preprocess are not well-aligned, the mapper / processor can be implemented in the VLLM model class as a backup plan to fix it

### Before submitting a new issue...

- [X] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

### PR DESCRIPTION
**Pull request by alex-jw-brooks**
FIX https://github.com/vllm-project/vllm/issues/8742

This PR adds support for allowing `mm_processor_kwargs` provided at inference time. This is a follow-up to the PR which added support for them at initialization time: https://github.com/vllm-project/vllm/pull/8657; with this, we can pass a `mm_processor_kwargs` dict at inference time, whose keys take precendence over the init time `mm_processor_kwargs` (which takes precedence over values pulled out of the HF config). 

Support is added for:
- Offline generation 
- Offline chat
for both decoder & encoder/decoder models.

It's not yet exposed through the server - I'm looking into exposing it there also, but thought this PR is already touching a lot of things and would be better to split up!

Here are a few examples for how it may be used:

1. `.generate()` with Phi3v `num_crops` (uses default mapper)
```python
from vllm import LLM, SamplingParams
from vllm.assets.image import ImageAsset

question = "What is the content of this image?"
prompt = f"<|user|>\n<|image_1|>\n{question}<|end|>\n<|assistant|>\n"
image = ImageAsset("cherry_blossom").pil_image.convert("RGB")

llm = LLM(
    model="microsoft/Phi-3-vision-128k-instruct",
    trust_remote_code=True,
    max_num_seqs=5,
    # Makes the defaults for num_crops 4 for this instance
    mm_processor_kwargs={"num_crops": 4},
    max_model_len=4096
)

sampling_params = SamplingParams(temperature=0, max_tokens=64)

outputs = llm.generate(
    {
        "prompt": prompt,
        "multi_modal_data": {
            "image": image
        },
        # Takes priority of the 4 we passed earlier; overrides with num_crops=2
        "mm_processor_kwargs": {"num_crops": 2}
    }, 
    sampling_params=sampling_params
)

for o in outputs:
    generated_text = o.outputs[0].text
    print(generated_text)
```

Example generate outputs (temperature=`0`):
`num_crops=2`: `The image captures a serene scene of a cityscape. Dominating the background is a tall, white tower with a pointed top, standing majestically against the backdrop of a clear blue sky. The tower's white color contrasts beautifully with the blue sky, creating a visually pleasing`

`num_crops=4`: ` The image captures a serene scene of a cityscape. Dominating the background is the iconic Space Needle, a tall, white tower with a gold top, standing majestically against the backdrop of a clear blue sky. The tower is partially obscured by a cluster of pink cherry`

2. `.chat()` with InternVL `max_dynamic_patch` (uses InternVL's custom mapper)
```python
from transformers import AutoTokenizer
from vllm import LLM, SamplingParams
from vllm.assets.image import ImageAsset
from vllm.multimodal.utils import fetch_image

question = "What is the content of this image?"
image_url = "https://upload.wikimedia.org/wikipedia/commons/d/da/2015_Kaczka_krzy%C5%BCowka_w_wodzie_%28samiec%29.jpg"
image_data = fetch_image(image_url)


model_name = "OpenGVLab/InternVL2-2B"
llm = LLM(
    model=model_name,
    trust_remote_code=True,
    max_model_len=4096,
    mm_processor_kwargs={"max_dynamic_patch": 4},
)
messages = [{'role': 'user', 'content': f"Image-1: <image>\n\n{question}"}]

tokenizer = AutoTokenizer.from_pretrained(model_name,
                                          trust_remote_code=True)
prompt = tokenizer.apply_chat_template(messages,
                                       tokenize=False,
                                       add_generation_prompt=True)

stop_tokens = ["<|endoftext|>", "<|im_start|>", "<|im_end|>", "<|end|>"]
stop_token_ids = [tokenizer.convert_tokens_to_ids(i) for i in stop_tokens]

sampling_params = SamplingParams(temperature=0,
                                max_tokens=128,
                                stop_token_ids=stop_token_ids)

outputs = llm.chat(
    [{
        "role":
        "user",
        "content": [
            {
                "type": "text",
                "text": question,
            },
            {
                "type": "image_url",
                "image_url": {
                    "url": image_url
                },
            },
        ],
    }],
    sampling_params=sampling_params,
    mm_processor_kwargs={"max_dynamic_patch": 8},
)

for o in outputs:
    generated_text = o.outputs[0].text
    print(generated_text)
```

Example chat outputs (temperature=`0`):
`max_dynamic_patch=4`: `The image shows a duck floating on water. The duck has a green head and neck, a white body, and a yellow beak. The water is calm and reflects the duck's image.`

`max_dynamic_patch=8`: `The image shows a duck floating on water. The duck has a green head and neck, a white body, and a yellow beak. The water is calm and blue, and there is a reflection of the duck visible on the water's surface.`

3. For encoder / decoder; I don't think there are any `mm_processor_kwargs` exposed yet. I did run the `offline_inference_encoder_decoder` (bart-large) demo to ensure that we see warnings coming out for unused `mm_processor_kwargs`, since the input processor happens for non-multimodal models also, and to test the synchronous input processing stuff a bit though, so hopefully between that and the two above ^ it tests mosts of the code-path for multimodal encoder/decoder models that will use `mm_processor_kwargs also`.

### Other Details
There is a bit of strangeness in implementing this, specifically when considering multiple modalities, because a model supporting multiple modalities has only one `input_processor`, but will have multiple mappers, since these are implemented per modality.

Because of this, it makes it a little strange to think through whether the API should be something like this (what's in the issue):

```python
llm.generate({"multi_modal_data": {"image": {"data": image, "mm_processor_kwargs": mm_processor_kwargs}}})
```

or this (what's implemented here):
```python
llm.generate({"multi_modal_data": {"image": image}}, "mm_processor_kwargs": mm_processor_kwargs)
```

I think there are arguments to be made for both, but after implementing it, the reason I wrote it as the second is as follows:
1. Adding support for the first one would add a lot of boiler plate to have models check `multi_modal_data[<modality>]` to see if it's an image type, or if it needs to be unwrapped, inside of pretty much all of our multimodal models

2. For the first one, users need to (sort of) know the separation of the `input_processor` & `mapper`; from the perspective of the input processor, it only makes sense if the per-modality keys are disjoint in their keys or duplicating data; e.g., if we had some thing like 

```python
llm.generate({"multi_modal_data": {
        "image": {"data": image, "mm_processor_kwargs": {"num_crops": 4}},
        "image": {"data": video, "mm_processor_kwargs": {"num_crops": 16}}
    }
})
```

It's fine in the mappers since they are handled per modality, but the input processor doesn't have a good way for easily understanding these values, at least in my opinion.

3. Because of that ^ I think it makes more sense to implement it in this format, so that the input processors and all mappers get the same mm_processor_kwargs. This also doesn't require changing any models, and aligns the APIs well for initialization time `mm_processor_kwargs` and inference time.

**BEFORE SUBMITTING, PLEASE READ THE CHECKLIST BELOW AND FILL IN THE DESCRIPTION ABOVE**

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Adding or changing kernels</h3>
<p>Each custom kernel needs a schema and one or more implementations to be registered with PyTorch.</p>
<ul>
    <li>Make sure custom ops are registered following PyTorch guidelines: <a href="https://pytorch.org/tutorials/advanced/cpp_custom_ops.html#cpp-custom-ops-tutorial">Custom C++ and CUDA Operators</a> and <a href="https://docs.google.com/document/d/1_W62p8WJOQQUzPsJYa7s701JXt0qf2OfLub2sbkHOaU">The Custom Operators Manual</a></li>
    <li>Custom operations that return <code>Tensors</code> require meta-functions. Meta-functions should be implemented and registered in python so that dynamic dims can be handled automatically. See above documents for a description of meta-functions.</li>
    <li>Use <a href="https://pytorch.org/docs/stable/library.html#torch.library.opcheck"><code>torch.libary.opcheck()</code></a> to test the function registration and meta-function for any registered ops.  See <code>tests/kernels</code> for examples.</li>
    <li>When changing the C++ signature of an existing op, the schema must be updated to reflect the changes.</li>
    <li>If a new custom type is needed, see the following document: <a href="https://docs.google.com/document/d/18fBMPuOJ0fY5ZQ6YyrHUppw9FA332CpNtgB6SOIgyuA">Custom Class Support in PT2</a>.
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

File: tests/multimodal/test_processor_kwargs.py [hunk 1]
```diff
@@ -271,32 +299,34 @@ def test_default_mapper_with_processer_kwargs(image_assets, num_crops):
...
-            mm_registry._get_plugin("image"),
-            "_default_input_mapper",
-        {mm_model_cls(): custom_mapper},
-    ):
-        mapped_inputs = mm_registry.map_input(ctx.model_config, mm_inputs)
+    # Patch the image registry for phi3v with our lambda that is compatible
+    # with overrides, then ensure that calling the method correctly echos
+    # our num_crops value back from the mm_processor_kwargs.
+    mm_registry._get_plugin("image").register_input_mapper(custom_mapper)(
+        mm_model_cls())
```
Comment (alex-jw-brooks) [inline tests/multimodal/test_processor_kwargs.py:327]: This was a bug in the previous test - the patching wasn't properly clobbering the input mapper with the custom mapper for testing, so it was still falling back to the default mapper here

**Review by DarkLight1337 — COMMENTED** [review]
> or this (what's implemented here):
> ```
> llm.generate({"multi_modal_data": {"image": image}}, "mm_processor_kwargs": > mm_processor_kwargs)
> ```

Agreed. @ywang96 and I have been thinking of combining tokenizer and processor together since that's how it's usually done in HuggingFace (where processor handles text + various modalities together), so we won't run into the issue of having to re-implement just the image/video processor part inside vLLM input processor. This implementation would tie better into our plan.

File: tests/test_inputs.py [hunk 2]
```diff
@@ -51,3 +52,40 @@ def test_parse_single_batch_token_consistent(token_input: List[int]):
...
+    ([{
+        "foo": 100
+    }, {
+        "bar": 200
+    }], [{
+        "foo": 100
+    }, {
+        "bar": 200
+    }]),
+])
```
[the remark this answers is not in the document]
Comment (alex-jw-brooks) [reply tests/test_inputs.py]: Yikes! Just did, nice catch 😄

File: vllm/utils.py [hunk 3]
```diff
@@ -1310,6 +1338,11 @@ def get_allowed_kwarg_only_overrides(
     if not overrides:
         return {}
 
+    # In some situations, the real callable might be wrapped, e.g., the init of
+    # a class received through a HF auto class. In such cases, allow anything.
+    if callable is None:
+        return overrides
```
[the remark this answers is not in the document]
Comment (alex-jw-brooks) [reply vllm/utils.py]: I think that's a great idea, this felt like a kind of gross hack to me 😬 I added some args to `supports_kw` for whether or not the args are allowed to be positional or not, and added some handling for weird special cases with variadic kwargs that may cause problems, then rewrote the `mm_processor_kwargs` handling stuff to leverage `supports_kw` internally 🤞

**Review by DarkLight1337 — APPROVED** [review]
Tests pass, so it should be ok. Thanks for implementing this!

**Comment by DarkLight1337** [comment]
See #13533

**Pull request merged**